internal/pkg/api/handleDownload.go

```diff
@@ -0,0 +1,387 @@
+func (d *Downloader) getPackage(requestPath string) error {
+	// sanity check file path
+	filePath := filepath.Clean(filepath.Join(d.root, requestPath))
+	if !strings.HasPrefix(filePath, d.root) {
+		return fmt.Errorf("filepath invalid %q is not in package cache %q", filePath, d.root)
+	}
+
+	url, err := d.upstreamURI.Parse(path.Join("downloads", requestPath))
+	if err != nil {
+		return fmt.Errorf("unable to build url from %q: %w", requestPath, err)
+	}
+
+	err = d.getFile(url.String(), filePath)
+	if err != nil {
+		return err
+	}
+
+	// verify download (using sha and asc)
+	if !strings.HasSuffix(requestPath, ".asc") && !strings.HasSuffix(requestPath, ".sha512") {
+		err = d.getFile(url.String()+".asc", filePath+".asc")
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return err
+		}
+		err = d.getFile(url.String()+".sha512", filePath+".sha512")
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return err
+		}
+		err = d.verifier.Verify(filePath)
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return fmt.Errorf("package verification failed: %w", err)
+		}
+	}
+
+	// delete package later
+	if d.retention > 0 {
+		go d.startDeletionTimer(d.retention, filePath)
+	}
+
+	return nil
+}
+
+// startDeletionTimer will call removeFile on path after dur
+func (d *Downloader) startDeletionTimer(dur time.Duration, path string) {
+	t := time.NewTimer(dur)
+	select {
+	case <-t.C:
+		if err := d.removeFile(path); err != nil {
+			log.Error().Err(err).Msg("unable to remove file")
+		}
+	case <-d.ctx.Done():
+		t.Stop()
+	}
+}
+
+// removeFile wraps remove with the filesystem lock
+func (d *Downloader) removeFile(path string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	return remove(path)
+}
+
+// remove will remove path ignoring any ErrNotExist errors
+// If path does not end in .asc or .sha512, then the .asc and .sha512 files are removed as well
+func remove(path string) error {
+	err := os.Remove(path)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return err
+	}
+	if !strings.HasSuffix(path, ".asc") && !strings.HasSuffix(path, ".sha512") {
+		if err := remove(path + ".asc"); err != nil {
+			return err
+		}
+		if err := remove(path + ".sha512"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getFile retrieves the file at url and writes the body to path.
+// If path contains subdirs they are created as well.
+// The file is opened with create/write/truncate flags.
+func (d *Downloader) getFile(url, path string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	rErr := &RetrievalError{url: url}
+
+	// TODO maybe check if the path exists?
+	err := os.MkdirAll(filepath.Dir(path), 0750)
+	if err != nil && !errors.Is(err, fs.ErrExist) {
```

Comment:
not sure if MkdirAll will ever return fs.ErrExist, maybe can remove ```&& !errors.Is(err, fs.ErrExist)```


```If path is already a directory, MkdirAll does nothing and returns nil.```
https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/os/path.go;l=18